Bottom: e0c238f79cebdf269b6052344ddf39e315b85648
Top:    0ea35aa8a27adcc3c9c0e563f70f3f6b13625a74
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-01-07 02:18:58 +0000

Move some fields from the HID* structure to HID_DRAW* and HID_DRAW_CLASS*

The parameters gui, poly_before, poly_after only effect rendering, so
belong in either the HID_DRAW_CLASS or HID_DRAW instance structure.

The set_layer() and end_layer() APIs also belong in the HID_DRAW_CLASS

We leave a copy of the "gui" flag in the HID structure, as main.c uses
this to count up each type of exporter.

XXX: THIS BREAKS SOME EXPORERS WHICH RELY ON THE NO-OP end_layer function
previously initialised in common_nogui_init()


---

diff --git a/src/draw.c b/src/draw.c
index 3083038..f9fc2c7 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -176,10 +176,10 @@ _draw_pv_name (PinType *pv)
   text.Y = box.Y1;
   text.Direction = vert ? 1 : 0;
 
-  if (gui->gui)
+  if (hid_draw_is_gui (gui->graphics))
     doing_pinout++;
   hid_draw_pcb_text (Output.fgGC, &text, 0);
-  if (gui->gui)
+  if (hid_draw_is_gui (gui->graphics))
     doing_pinout--;
 }
 
@@ -320,7 +320,7 @@ pad_callback (const BoxType * b, void *cl)
 static void
 draw_element_name (ElementType *element)
 {
-  if ((TEST_FLAG (HIDENAMESFLAG, PCB) && gui->gui) ||
+  if ((TEST_FLAG (HIDENAMESFLAG, PCB) && hid_draw_is_gui (gui->graphics)) ||
       TEST_FLAG (HIDENAMEFLAG, element))
     return;
   if (doing_pinout || doing_assy)
@@ -566,7 +566,7 @@ DrawEverything (const BoxType *drawn_area)
    * first draw all 'invisible' stuff
    */
   if (!TEST_FLAG (CHECKPLANESFLAG, PCB)
-      && gui->set_layer ("invisible", SL (INVISIBLE, 0), 0))
+      && hid_draw_set_layer (gui->graphics, "invisible", SL (INVISIBLE, 0), 0))
     {
       side = SWAP_IDENT ? TOP_SIDE : BOTTOM_SIDE;
       if (PCB->ElementOn)
@@ -576,7 +576,7 @@ DrawEverything (const BoxType *drawn_area)
 	  DrawLayer (&(PCB->Data->Layer[max_copper_layer + side]), drawn_area);
 	}
       r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
   /* draw all layers in layerstack order */
@@ -584,105 +584,105 @@ DrawEverything (const BoxType *drawn_area)
     {
       int group = drawn_groups[i];
 
-      if (gui->set_layer (0, group, 0))
+      if (hid_draw_set_layer (gui->graphics, 0, group, 0))
         {
           DrawLayerGroup (group, drawn_area);
-          gui->end_layer ();
+          hid_draw_end_layer (gui->graphics);
         }
     }
 
-  if (TEST_FLAG (CHECKPLANESFLAG, PCB) && gui->gui)
+  if (TEST_FLAG (CHECKPLANESFLAG, PCB) && hid_draw_is_gui (gui->graphics))
     return;
 
   /* Draw pins, pads, vias below silk */
-  if (gui->gui)
+  if (hid_draw_is_gui (gui->graphics))
     DrawPPV (SWAP_IDENT ? bottom_group : top_group, drawn_area);
   else
     {
       CountHoles (&plated, &unplated, drawn_area);
 
-      if (plated && gui->set_layer ("plated-drill", SL (PDRILL, 0), 0))
+      if (plated && hid_draw_set_layer (gui->graphics, "plated-drill", SL (PDRILL, 0), 0))
         {
           DrawHoles (true, false, drawn_area);
-          gui->end_layer ();
+          hid_draw_end_layer (gui->graphics);
         }
 
-      if (unplated && gui->set_layer ("unplated-drill", SL (UDRILL, 0), 0))
+      if (unplated && hid_draw_set_layer (gui->graphics, "unplated-drill", SL (UDRILL, 0), 0))
         {
           DrawHoles (false, true, drawn_area);
-          gui->end_layer ();
+          hid_draw_end_layer (gui->graphics);
         }
     }
 
   /* Draw the solder mask if turned on */
-  if (gui->set_layer ("componentmask", SL (MASK, TOP), 0))
+  if (hid_draw_set_layer (gui->graphics, "componentmask", SL (MASK, TOP), 0))
     {
       DrawMask (TOP_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
-  if (gui->set_layer ("soldermask", SL (MASK, BOTTOM), 0))
+  if (hid_draw_set_layer (gui->graphics, "soldermask", SL (MASK, BOTTOM), 0))
     {
       DrawMask (BOTTOM_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
-  if (gui->set_layer ("topsilk", SL (SILK, TOP), 0))
+  if (hid_draw_set_layer (gui->graphics, "topsilk", SL (SILK, TOP), 0))
     {
       DrawSilk (TOP_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
-  if (gui->set_layer ("bottomsilk", SL (SILK, BOTTOM), 0))
+  if (hid_draw_set_layer (gui->graphics, "bottomsilk", SL (SILK, BOTTOM), 0))
     {
       DrawSilk (BOTTOM_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
-  if (gui->gui)
+  if (hid_draw_is_gui (gui->graphics))
     {
       /* Draw element Marks */
       if (PCB->PinOn)
 	r_search (PCB->Data->element_tree, drawn_area, NULL, EMark_callback,
 		  NULL);
       /* Draw rat lines on top */
-      if (gui->set_layer ("rats", SL (RATS, 0), 0))
+      if (hid_draw_set_layer (gui->graphics, "rats", SL (RATS, 0), 0))
         {
           DrawRats(drawn_area);
-          gui->end_layer ();
+          hid_draw_end_layer (gui->graphics);
         }
     }
 
   paste_empty = IsPasteEmpty (TOP_SIDE);
-  if (gui->set_layer ("toppaste", SL (PASTE, TOP), paste_empty))
+  if (hid_draw_set_layer (gui->graphics, "toppaste", SL (PASTE, TOP), paste_empty))
     {
       DrawPaste (TOP_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
   paste_empty = IsPasteEmpty (BOTTOM_SIDE);
-  if (gui->set_layer ("bottompaste", SL (PASTE, BOTTOM), paste_empty))
+  if (hid_draw_set_layer (gui->graphics, "bottompaste", SL (PASTE, BOTTOM), paste_empty))
     {
       DrawPaste (BOTTOM_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
-  if (gui->set_layer ("topassembly", SL (ASSY, TOP), 0))
+  if (hid_draw_set_layer (gui->graphics, "topassembly", SL (ASSY, TOP), 0))
     {
       PrintAssembly (TOP_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
-  if (gui->set_layer ("bottomassembly", SL (ASSY, BOTTOM), 0))
+  if (hid_draw_set_layer (gui->graphics, "bottomassembly", SL (ASSY, BOTTOM), 0))
     {
       PrintAssembly (BOTTOM_SIDE, drawn_area);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 
-  if (gui->set_layer ("fab", SL (FAB, 0), 0))
+  if (hid_draw_set_layer (gui->graphics, "fab", SL (FAB, 0), 0))
     {
       PrintFab (Output.fgGC);
-      gui->end_layer ();
+      hid_draw_end_layer (gui->graphics);
     }
 }
 
@@ -740,7 +740,7 @@ DrawPPV (int group, const BoxType *drawn_area)
   int bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
   int side;
 
-  if (PCB->PinOn || !gui->gui)
+  if (PCB->PinOn || !hid_draw_is_gui (gui->graphics))
     {
       /* draw element pins */
       r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
@@ -760,7 +760,7 @@ DrawPPV (int group, const BoxType *drawn_area)
     }
 
   /* draw vias */
-  if (PCB->ViaOn || !gui->gui)
+  if (PCB->ViaOn || !hid_draw_is_gui (gui->graphics))
     {
       r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
       r_search (PCB->Data->via_tree, drawn_area, NULL, hole_callback, NULL);
@@ -822,7 +822,7 @@ DrawSilk (int side, const BoxType * drawn_area)
 #endif
 
 #if 0
-  if (gui->poly_before)
+  if (gui->graphics->poly_before)
     {
       hid_draw_use_mask (gui->graphics, HID_MASK_BEFORE);
 #endif
@@ -838,7 +838,7 @@ DrawSilk (int side, const BoxType * drawn_area)
   r_search (PCB->Data->via_tree, drawn_area, NULL, clearPin_callback, NULL);
   r_search (PCB->Data->pad_tree, drawn_area, NULL, clearPad_callback, &side);
 
-  if (gui->poly_after)
+  if (gui->graphics->poly_after)
     {
       hid_draw_use_mask (gui->graphics, HID_MASK_AFTER);
       DrawLayer (LAYER_PTR (max_copper_layer + layer), drawn_area);
@@ -855,8 +855,8 @@ static void
 DrawMaskBoardArea (int mask_type, const BoxType *drawn_area)
 {
   /* Skip the mask drawing if the GUI doesn't want this type */
-  if ((mask_type == HID_MASK_BEFORE && !gui->poly_before) ||
-      (mask_type == HID_MASK_AFTER  && !gui->poly_after))
+  if ((mask_type == HID_MASK_BEFORE && !gui->graphics->poly_before) ||
+      (mask_type == HID_MASK_AFTER  && !gui->graphics->poly_after))
     return;
 
   hid_draw_use_mask (gui->graphics, mask_type);
@@ -974,7 +974,7 @@ DrawLayer (LayerType *Layer, const BoxType *screen)
   /* draw the layer text on screen */
   r_search (Layer->text_tree, screen, NULL, text_callback, Layer);
 
-  /* We should check for gui->gui here, but it's kinda cool seeing the
+  /* We should check for hid_draw_is_gui (gui->graphics) here, but it's kinda cool seeing the
      auto-outline magically disappear when you first add something to
      the "outline" layer.  */
   if (IsLayerEmpty (Layer)
@@ -1013,7 +1013,7 @@ DrawLayerGroup (int group, const BoxType *drawn_area)
   if (n_entries > 1)
     rv = 1;
 
-  if (rv && !gui->gui)
+  if (rv && !hid_draw_is_gui (gui->graphics))
     DrawPPV (group, drawn_area);
 }
 
diff --git a/src/hid.h b/src/hid.h
index 5ff0c2c..26d7edc 100644
--- a/src/hid.h
+++ b/src/hid.h
@@ -269,15 +269,6 @@ typedef enum
        and EPS exporters.  */
     char exporter:1;
 
-    /* If set, the redraw code will draw polygons before erasing the
-       clearances.  */
-    char poly_before:1;
-
-    /* If set, the redraw code will draw polygons after erasing the
-       clearances.  Note that HIDs may set both of these, in which case
-       polygons will be drawn twice.  */
-    char poly_after:1;
-
     /* Returns a set of resources describing options the export or print
        HID supports.  In GUI mode, the print/export dialogs use this to
        set up the selectable options.  In command line mode, these are
@@ -303,23 +294,6 @@ typedef enum
     void (*notify_crosshair_change) (bool changes_complete);
     void (*notify_mark_change) (bool changes_complete);
 
-    /* During redraw or print/export cycles, this is called once per
-       layer (or layer group, for copper layers).  If it returns false
-       (zero), the HID does not want that layer, and none of the drawing
-       functions should be called.  If it returns true (nonzero), the
-       items in that layer [group] should be drawn using the various
-       drawing functions.  In addition to the MAX_GROUP copper layer
-       groups, you may select layers indicated by the macros SL_*
-       defined above, or any others with an index of -1.  For copper
-       layer groups, you may pass NULL for name to have a name fetched
-       from the PCB struct.  The EMPTY argument is a hint - if set, the
-       layer is empty, if zero it may be non-empty.  */
-    int (*set_layer) (const char *name_, int group_, int _empty);
-
-    /* Tell the GUI the layer last selected has been finished with */
-    void (*end_layer) (void);
-
-
     HID_DRAW *graphics;
 
     /* This is for the printer.  If you call this for the GUI, xval and
diff --git a/src/hid/batch/batch.c b/src/hid/batch/batch.c
index 252e2bf..4f0acb3 100644
--- a/src/hid/batch/batch.c
+++ b/src/hid/batch/batch.c
@@ -344,7 +344,6 @@ hid_batch_init ()
   batch_hid.parse_arguments       = batch_parse_arguments;
   batch_hid.invalidate_lr         = batch_invalidate_lr;
   batch_hid.invalidate_all        = batch_invalidate_all;
-  batch_hid.set_layer             = batch_set_layer;
   batch_hid.calibrate             = batch_calibrate;
   batch_hid.shift_is_pressed      = batch_shift_is_pressed;
   batch_hid.control_is_pressed    = batch_control_is_pressed;
@@ -364,6 +363,7 @@ hid_batch_init ()
 
   common_draw_helpers_class_init (&batch_graphics_class);
 
+  batch_graphics_class.set_layer      = batch_set_layer;
   batch_graphics_class.make_gc        = batch_make_gc;
   batch_graphics_class.destroy_gc     = batch_destroy_gc;
   batch_graphics_class.use_mask       = batch_use_mask;
diff --git a/src/hid/common/extents.c b/src/hid/common/extents.c
index cf2de59..1b4746f 100644
--- a/src/hid/common/extents.c
+++ b/src/hid/common/extents.c
@@ -181,14 +181,12 @@ hid_extents_init (void)
   extents_hid.struct_size         = sizeof (HID);
   extents_hid.name                = "extents-extents";
   extents_hid.description         = "used to calculate extents";
-  extents_hid.poly_before         = 1;
-
-  extents_hid.set_layer           = extents_set_layer;
 
   extents_hid.graphics            = &extents_graphics;
 
   common_draw_helpers_class_init (&extents_graphics_class);
 
+  extents_graphics_class.set_layer      = extents_set_layer;
   extents_graphics_class.make_gc        = extents_make_gc;
   extents_graphics_class.destroy_gc     = extents_destroy_gc;
   extents_graphics_class.use_mask       = extents_use_mask;
@@ -204,6 +202,7 @@ hid_extents_init (void)
   extents_graphics_class.fill_rect      = extents_fill_rect;
 
   extents_graphics.klass = &extents_graphics_class;
+  extents_graphics.poly_before = true;
   common_draw_helpers_init (&extents_graphics);
 
   initialised = true;
diff --git a/src/hid/common/hidnogui.c b/src/hid/common/hidnogui.c
index 3b85044..85e678d 100644
--- a/src/hid/common/hidnogui.c
+++ b/src/hid/common/hidnogui.c
@@ -448,8 +448,6 @@ common_nogui_init (HID *hid)
   hid->parse_arguments =      nogui_parse_arguments;
   hid->invalidate_lr =        nogui_invalidate_lr;
   hid->invalidate_all =       nogui_invalidate_all;
-  hid->set_layer =            nogui_set_layer;
-  hid->end_layer =            nogui_end_layer;
   hid->calibrate =            nogui_calibrate;
   hid->shift_is_pressed =     nogui_shift_is_pressed;
   hid->control_is_pressed =   nogui_control_is_pressed;
@@ -481,6 +479,8 @@ common_nogui_init (HID *hid)
 static void
 common_nogui_graphics_class_init (HID_DRAW_CLASS *klass)
 {
+  klass->set_layer =       nogui_set_layer;
+  klass->end_layer =       nogui_end_layer;
   klass->make_gc =         nogui_make_gc;
   klass->destroy_gc =      nogui_destroy_gc;
   klass->use_mask =        nogui_use_mask;
diff --git a/src/hid/gcode/gcode.c b/src/hid/gcode/gcode.c
index dc3ffcf..9fca3c1 100644
--- a/src/hid/gcode/gcode.c
+++ b/src/hid/gcode/gcode.c
@@ -1619,12 +1619,10 @@ hid_gcode_init ()
   gcode_hid.name                = "gcode";
   gcode_hid.description         = "G-CODE export";
   gcode_hid.exporter            = 1;
-  gcode_hid.poly_before         = 1;
 
   gcode_hid.get_export_options  = gcode_get_export_options;
   gcode_hid.do_export           = gcode_do_export;
   gcode_hid.parse_arguments     = gcode_parse_arguments;
-  gcode_hid.set_layer           = gcode_set_layer;
   gcode_hid.calibrate           = gcode_calibrate;
   gcode_hid.set_crosshair       = gcode_set_crosshair;
 
@@ -1632,6 +1630,7 @@ hid_gcode_init ()
 
   common_draw_helpers_class_init (&gcode_graphics_class);
 
+  gcode_graphics_class.set_layer      = gcode_set_layer;
   gcode_graphics_class.make_gc        = gcode_make_gc;
   gcode_graphics_class.destroy_gc     = gcode_destroy_gc;
   gcode_graphics_class.use_mask       = gcode_use_mask;
@@ -1648,6 +1647,7 @@ hid_gcode_init ()
   gcode_graphics_class.fill_rect      = gcode_fill_rect;
 
   gcode_graphics.klass = &gcode_graphics_class;
+  gcode_graphics.poly_before = true;
   common_draw_helpers_init (&gcode_graphics);
 
   hid_register_hid (&gcode_hid);
diff --git a/src/hid/gerber/gerber.c b/src/hid/gerber/gerber.c
index 6185304..22ae4d6 100644
--- a/src/hid/gerber/gerber.c
+++ b/src/hid/gerber/gerber.c
@@ -1355,7 +1355,6 @@ hid_gerber_init ()
   gerber_hid.get_export_options  = gerber_get_export_options;
   gerber_hid.do_export           = gerber_do_export;
   gerber_hid.parse_arguments     = gerber_parse_arguments;
-  gerber_hid.set_layer           = gerber_set_layer;
   gerber_hid.calibrate           = gerber_calibrate;
   gerber_hid.set_crosshair       = gerber_set_crosshair;
 
@@ -1363,6 +1362,7 @@ hid_gerber_init ()
 
   common_draw_helpers_class_init (&gerber_graphics_class);
 
+  gerber_graphics_class.set_layer      = gerber_set_layer;
   gerber_graphics_class.make_gc        = gerber_make_gc;
   gerber_graphics_class.destroy_gc     = gerber_destroy_gc;
   gerber_graphics_class.use_mask       = gerber_use_mask;
diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index b5e4a00..4690dd5 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -803,7 +803,7 @@ ghid_init_renderer (int *argc, char ***argv, GHidPort *port)
     }
 
   /* Setup HID function pointers specific to the GL renderer*/
-  ghid_hid.end_layer = ghid_end_layer;
+  ghid_graphics_class.end_layer = ghid_end_layer;
   ghid_graphics_class.fill_pcb_polygon = ghid_fill_pcb_polygon;
   ghid_graphics_class.thindraw_pcb_polygon = ghid_thindraw_pcb_polygon;
 }
diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index 1a64566..c44b7e1 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -2140,7 +2140,6 @@ hid_gtk_init ()
   ghid_hid.name                     = "gtk";
   ghid_hid.description              = "Gtk - The Gimp Toolkit";
   ghid_hid.gui                      = 1;
-  ghid_hid.poly_after               = 1;
 
   ghid_hid.get_export_options       = ghid_get_export_options;
   ghid_hid.do_export                = ghid_do_export;
@@ -2149,7 +2148,6 @@ hid_gtk_init ()
   ghid_hid.invalidate_all           = ghid_invalidate_all;
   ghid_hid.notify_crosshair_change  = ghid_notify_crosshair_change;
   ghid_hid.notify_mark_change       = ghid_notify_mark_change;
-  ghid_hid.set_layer                = ghid_set_layer;
 
   ghid_hid.calibrate                = ghid_calibrate;
   ghid_hid.shift_is_pressed         = ghid_shift_is_pressed;
@@ -2189,6 +2187,7 @@ hid_gtk_init ()
 
   common_draw_helpers_class_init (&ghid_graphics_class);
 
+  ghid_graphics_class.set_layer      = ghid_set_layer;
   ghid_graphics_class.make_gc        = ghid_make_gc;
   ghid_graphics_class.destroy_gc     = ghid_destroy_gc;
   ghid_graphics_class.use_mask       = ghid_use_mask;
@@ -2205,7 +2204,10 @@ hid_gtk_init ()
 
   ghid_graphics_class.draw_pcb_polygon = common_gui_draw_pcb_polygon;
 
+  ghid_graphics_class.gui = true;
+
   ghid_graphics.klass = &ghid_graphics_class;
+  ghid_graphics.poly_after = true;
   common_draw_helpers_init (&ghid_graphics);
 
   hid_register_hid (&ghid_hid);
diff --git a/src/hid/lesstif/main.c b/src/hid/lesstif/main.c
index fbfcc7f..698dd63 100644
--- a/src/hid/lesstif/main.c
+++ b/src/hid/lesstif/main.c
@@ -4076,7 +4076,6 @@ hid_lesstif_init ()
   lesstif_hid.name                    = "lesstif";
   lesstif_hid.description             = "LessTif - a Motif clone for X/Unix";
   lesstif_hid.gui                     = 1;
-  lesstif_hid.poly_before             = 1;
 
   lesstif_hid.get_export_options      = lesstif_get_export_options;
   lesstif_hid.do_export               = lesstif_do_export;
@@ -4085,7 +4084,6 @@ hid_lesstif_init ()
   lesstif_hid.invalidate_all          = lesstif_invalidate_all;
   lesstif_hid.notify_crosshair_change = lesstif_notify_crosshair_change;
   lesstif_hid.notify_mark_change      = lesstif_notify_mark_change;
-  lesstif_hid.set_layer               = lesstif_set_layer;
 
   lesstif_hid.calibrate               = lesstif_calibrate;
   lesstif_hid.shift_is_pressed        = lesstif_shift_is_pressed;
@@ -4121,6 +4119,7 @@ hid_lesstif_init ()
 
   common_draw_helpers_class_init (&lesstif_graphics_class);
 
+  lesstif_graphics_class.set_layer      = lesstif_set_layer;
   lesstif_graphics_class.make_gc        = lesstif_make_gc;
   lesstif_graphics_class.destroy_gc     = lesstif_destroy_gc;
   lesstif_graphics_class.use_mask       = lesstif_use_mask;
@@ -4138,6 +4137,7 @@ hid_lesstif_init ()
   lesstif_graphics_class.draw_pcb_polygon = common_gui_draw_pcb_polygon;
 
   lesstif_graphics.klass = &lesstif_graphics_class;
+  lesstif_graphics.poly_before = true;
   common_draw_helpers_init (&lesstif_graphics);
 
   hid_register_hid (&lesstif_hid);
diff --git a/src/hid/lpr/lpr.c b/src/hid/lpr/lpr.c
index 05d5863..da9e20f 100644
--- a/src/hid/lpr/lpr.c
+++ b/src/hid/lpr/lpr.c
@@ -136,7 +136,6 @@ hid_lpr_init ()
   lpr_hid.name                = "lpr";
   lpr_hid.description         = N_("Postscript print");
   lpr_hid.printer             = 1;
-  lpr_hid.poly_before         = 1;
 
   lpr_hid.get_export_options  = lpr_get_export_options;
   lpr_hid.do_export           = lpr_do_export;
diff --git a/src/hid/nelma/nelma.c b/src/hid/nelma/nelma.c
index cb28849..dff74b0 100644
--- a/src/hid/nelma/nelma.c
+++ b/src/hid/nelma/nelma.c
@@ -1070,12 +1070,10 @@ hid_nelma_init()
   nelma_hid.name                = "nelma";
   nelma_hid.description         = "Numerical analysis package export";
   nelma_hid.exporter            = 1;
-  nelma_hid.poly_before         = 1;
 
   nelma_hid.get_export_options  = nelma_get_export_options;
   nelma_hid.do_export           = nelma_do_export;
   nelma_hid.parse_arguments     = nelma_parse_arguments;
-  nelma_hid.set_layer           = nelma_set_layer;
   nelma_hid.calibrate           = nelma_calibrate;
   nelma_hid.set_crosshair       = nelma_set_crosshair;
 
@@ -1083,6 +1081,7 @@ hid_nelma_init()
 
   common_draw_helpers_class_init (&nelma_graphics_class);
 
+  nelma_graphics_class.set_layer      = nelma_set_layer;
   nelma_graphics_class.make_gc        = nelma_make_gc;
   nelma_graphics_class.destroy_gc     = nelma_destroy_gc;
   nelma_graphics_class.use_mask       = nelma_use_mask;
@@ -1099,6 +1098,7 @@ hid_nelma_init()
   nelma_graphics_class.fill_rect      = nelma_fill_rect;
 
   nelma_graphics.klass = &nelma_graphics_class;
+  nelma_graphics.poly_before = true;
   common_draw_helpers_init (&nelma_graphics);
 
   hid_register_hid (&nelma_hid);
diff --git a/src/hid/png/png.c b/src/hid/png/png.c
index edb7c28..6ae0811 100644
--- a/src/hid/png/png.c
+++ b/src/hid/png/png.c
@@ -1845,12 +1845,10 @@ hid_png_init ()
   png_hid.name        = "png";
   png_hid.description = "GIF/JPEG/PNG export";
   png_hid.exporter    = 1;
-  png_hid.poly_before = 1;
 
   png_hid.get_export_options  = png_get_export_options;
   png_hid.do_export           = png_do_export;
   png_hid.parse_arguments     = png_parse_arguments;
-  png_hid.set_layer           = png_set_layer;
   png_hid.calibrate           = png_calibrate;
   png_hid.set_crosshair       = png_set_crosshair;
 
@@ -1858,6 +1856,7 @@ hid_png_init ()
 
   common_draw_helpers_class_init (&png_graphics_class);
 
+  png_graphics_class.set_layer      = png_set_layer;
   png_graphics_class.make_gc        = png_make_gc;
   png_graphics_class.destroy_gc     = png_destroy_gc;
   png_graphics_class.use_mask       = png_use_mask;
@@ -1873,6 +1872,7 @@ hid_png_init ()
   png_graphics_class.fill_rect      = png_fill_rect;
 
   png_graphics.klass = &png_graphics_class;
+  png_graphics.poly_before = true;
   common_draw_helpers_init (&png_graphics);
 
 #ifdef HAVE_SOME_FORMAT
diff --git a/src/hid/ps/eps.c b/src/hid/ps/eps.c
index 5890e4c..3f0afc6 100644
--- a/src/hid/ps/eps.c
+++ b/src/hid/ps/eps.c
@@ -240,11 +240,11 @@ eps_hid_export_to_file (FILE * the_file, HID_Attr_Val * options)
       print_layer[i] = 1;
 
   if (fast_erase) {
-    eps_hid.poly_before = 1;
-    eps_hid.poly_after = 0;
+    eps_graphics.poly_before = true;
+    eps_graphics.poly_after = false;
   } else {
-    eps_hid.poly_before = 0;
-    eps_hid.poly_after = 1;
+    eps_graphics.poly_before = false;
+    eps_graphics.poly_after = true;
   }
 
   memcpy (saved_layer_stack, LayerStack, sizeof (LayerStack));
@@ -686,12 +686,10 @@ hid_eps_init ()
   eps_hid.name                = "eps";
   eps_hid.description         = "Encapsulated Postscript";
   eps_hid.exporter            = 1;
-  eps_hid.poly_after          = 1;
 
   eps_hid.get_export_options  = eps_get_export_options;
   eps_hid.do_export           = eps_do_export;
   eps_hid.parse_arguments     = eps_parse_arguments;
-  eps_hid.set_layer           = eps_set_layer;
   eps_hid.calibrate           = eps_calibrate;
   eps_hid.set_crosshair       = eps_set_crosshair;
 
@@ -699,6 +697,7 @@ hid_eps_init ()
 
   common_draw_helpers_class_init (&eps_graphics_class);
 
+  eps_graphics_class.set_layer      = eps_set_layer;
   eps_graphics_class.make_gc        = eps_make_gc;
   eps_graphics_class.destroy_gc     = eps_destroy_gc;
   eps_graphics_class.use_mask       = eps_use_mask;
@@ -714,6 +713,7 @@ hid_eps_init ()
   eps_graphics_class.fill_rect      = eps_fill_rect;
 
   eps_graphics.klass = &eps_graphics_class;
+  eps_graphics.poly_after = true;
   common_draw_helpers_init (&eps_graphics);
 
   hid_register_hid (&eps_hid);
diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index a52bf76..1572ccc 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -1528,13 +1528,13 @@ void ps_ps_init (HID *hid)
   hid->get_export_options = ps_get_export_options;
   hid->do_export          = ps_do_export;
   hid->parse_arguments    = ps_parse_arguments;
-  hid->set_layer          = ps_set_layer;
   hid->calibrate          = ps_calibrate;
   hid->set_crosshair      = ps_set_crosshair;
 }
 
 void ps_ps_graphics_class_init (HID_DRAW_CLASS *klass)
 {
+  klass->set_layer          = ps_set_layer;
   klass->make_gc            = ps_make_gc;
   klass->destroy_gc         = ps_destroy_gc;
   klass->use_mask           = ps_use_mask;
@@ -1570,7 +1570,6 @@ hid_ps_init ()
   ps_hid.name               = "ps";
   ps_hid.description        = N_("Postscript export");
   ps_hid.exporter           = 1;
-  ps_hid.poly_before        = 1;
 
   ps_hid.graphics           = &ps_graphics;
 
@@ -1578,6 +1577,7 @@ hid_ps_init ()
   ps_ps_graphics_class_init (&ps_graphics_class);
 
   ps_graphics.klass = &ps_graphics_class;
+  ps_graphics.poly_before = true;
   common_draw_helpers_init (&ps_graphics);
   ps_ps_graphics_init (&ps_graphics);
 
diff --git a/src/hid_draw.h b/src/hid_draw.h
index ea288c7..a442390 100644
--- a/src/hid_draw.h
+++ b/src/hid_draw.h
@@ -9,6 +9,22 @@ enum mask_mode {
 
 typedef struct hid_draw_class_st
 {
+  /* During redraw or print/export cycles, this is called once per
+     layer (or layer group, for copper layers).  If it returns false
+     (zero), the HID does not want that layer, and none of the drawing
+     functions should be called.  If it returns true (nonzero), the
+     items in that layer [group] should be drawn using the various
+     drawing functions.  In addition to the MAX_GROUP copper layer
+     groups, you may select layers indicated by the macros SL_*
+     defined above, or any others with an index of -1.  For copper
+     layer groups, you may pass NULL for name to have a name fetched
+     from the PCB struct.  The EMPTY argument is a hint - if set, the
+     layer is empty, if zero it may be non-empty.  */
+  int (*set_layer) (const char *name_, int group_, int _empty);
+
+  /* Tell the GUI the layer last selected has been finished with */
+  void (*end_layer) (void);
+
   /* Drawing Functions.  Coordinates and distances are ALWAYS in PCB's
      default coordinates (1 nm at the time this comment was written).
      Angles are always in degrees, with 0 being "right" (positive X)
@@ -63,12 +79,19 @@ typedef struct hid_draw_class_st
   void (*fill_pcb_pv) (hidGC fg_gc, hidGC bg_gc, PinType *pv, bool drawHole, bool mask);
   void (*thindraw_pcb_pv) (hidGC fg_gc, hidGC bg_gc, PinType *pv, bool drawHole, bool mask);
 
+  /* draw.c currently uses this to shortcut and special-case certain rendering when displaying on-screen */
+  bool gui;
+
 } HID_DRAW_CLASS;
 
 /* Base HID_DRAW elements visible to any module */
 struct hid_draw_st
 {
   HID_DRAW_CLASS *klass;
+
+  /* Note that both of these may be set, in which case polygons will be drawn twice: */
+  bool poly_before; /* If set, the redraw code will draw polygons before erasing the clearances. */
+  bool poly_after;  /* If set, the redraw code will draw polygons after  erasing the clearances. */
 };
 
 /* Base hidGC elements visible to any module */
@@ -77,8 +100,29 @@ struct hid_gc_struct {
   HID_DRAW *hid_draw;
 };
 
+
+/* Getters for class properties */
+inline bool
+hid_draw_is_gui (HID_DRAW *hid_draw)
+{
+  return hid_draw->klass->gui;
+}
+
+
 /* Calling wrappers to access the vfunc table */
 
+inline int
+hid_draw_set_layer (HID_DRAW *hid_draw, const char *name, int group, int empty)
+{
+  return hid_draw->klass->set_layer (name, group, empty);
+}
+
+inline void
+hid_draw_end_layer (HID_DRAW *hid_draw)
+{
+  hid_draw->klass->end_layer ();
+}
+
 inline hidGC
 hid_draw_make_gc (HID_DRAW *hid_draw)
 {
